### PR TITLE
Never use templates when managing git repos

### LINKED ISSUE #6240: "failed to create temporary file" when building a dependency from Git
**Problem**
When trying to install a dependency from git I get the following error no matter which project I try to build if it has a git dependency.

```
$ cargo build
    Updating git repository `https://github.com/zacps/winpty`
warning: spurious network error (2 tries remaining): failed to create temporary file '/pack_git2_es7hEy': Permission denied; class=Os (2)
warning: spurious network error (1 tries remaining): failed to create temporary file '/pack_git2_rhZ8dQ': Permission denied; class=Os (2)
error: failed to load source for a dependency on `winpty-sys`

Caused by:
  Unable to update https://github.com/zacps/winpty?branch=rust#9ccec645

Caused by:
  failed to fetch into /Users/stanko/.cargo/git/db/winpty-788a6abf0b1935fd

Caused by:
  failed to create temporary file '/pack_git2_oFAHxn': Permission denied; class=Os (2)
```

**Steps**
1. Create a new project `cargo new --bin foobar`
2. Move into the project `cd foobar`
3. Add any git dependency to it e.g.  `echo "diesel = { git = \"https://github.com/diesel-rs/diesel\" }" >> Cargo.toml`
4. Run `cargo build`

**Possible Solution(s)**

By trying to run `cargo` as root it seems like `cargo` tries to build git dependencies inside the root directory, which is also indicated by the error as the temporary file is `/pack_git2_oFAHxn`. I'm unsure as to why this is happening.

**Notes**

I tried to reinstall Rust but it didn't resolve the issue.
I'm able to clone git repositories and use git.
I do have a `templatedir` set in my gitconfig (this was indicated to be the problem in a similar issue) but removing it from git, or removing it completely, or removing it from gitconfig didn't solve this issue.
I noticed this issue after updating from Rust 1.27 to 1.29, and now on 1.30.

Output of `cargo version`: `cargo 1.30.0 (36d96825d 2018-10-24)`
Output of `rustc --version`: `rustc 1.30.0 (da5f414c2 2018-10-24)`
Output of `rustup toolchain list`: `stable-x86_64-apple-darwin (default)`
Output of `git --version`: `git version 2.17.1 (Apple Git-112)`
OS: macOS Version 10.13.6 High Sierra

Contents of `~/.cargo`:
```
total 8
drwxr-xr-x    6 stanko  staff   192B Oct 30 08:57 .
drwxr-xr-x+ 112 stanko  staff   3.5K Oct 30 09:15 ..
drwxr-xr-x   12 stanko  staff   384B Oct 30 08:56 bin
-rw-r--r--    1 stanko  staff    37B Oct 30 08:56 env
drwxr-xr-x    4 stanko  staff   128B Oct 30 08:57 git
drwxr-xr-x    3 stanko  staff    96B Oct 30 08:57 registry
```

Output of `printenv`:

```
Apple_PubSub_Socket_Render=/private/tmp/com.apple.launchd.retracted/Render
BROWSER=/Applications/FirefoxDeveloperEdition.app/Contents/MacOS/firefox
EDITOR=nvim
ELIXIR_VERSION=1.6.4
ERL_AFLAGS=-kernel shell_history enabled
FZF_DEFAULT_COMMAND=rg --files --no-ignore --hidden --follow --glob "!.git/*"
GPG_TTY=/dev/ttys004
HOME=/Users/stanko
HOMEBREW_GITHUB_API_TOKEN=retracted
LANG=en_US.UTF-8
LC_CTYPE=en_US.UTF-8
LESS=-R
LOGNAME=stanko
LSCOLORS=Gxfxcxdxbxegedabagacad
MANPATH=/Users/stanko/.kerl/installs/20.3/lib/erlang/man:/Users/stanko/.kerl/installs/20.3/man:/usr/share/man:/usr/local/share/man:/usr/local/MacGPG2/share/man:/Users/stanko/.kerl/installs/20.3/lib/erlang/man:/Users/stanko/.kerl/installs/20.3/man
MIX_ARCHIVES=/Users/stanko/.kiex/mix/archives/elixir-1.6.4
OLDPWD=/Users/stanko
PAGER=less
PATH=/Users/stanko/.cargo/bin:/Users/stanko/.kiex/elixirs/elixir-1.6.4/bin:/Users/stanko/.kerl/installs/20.3/bin:/Users/stanko/.kiex/elixirs/elixir-1.6.4/bin:/Users/stanko/.kiex/bin:/Users/stanko/.rbenv/shims:/Users/stanko/.rbenv/bin:/Users/stanko/.rbenv/shims:/Users/stanko/.rbenv/bin:/Users/stanko/.cargo/bin:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/usr/local/MacGPG2/bin:/Users/stanko/.config/gcloud/google-cloud-sdk/bin:/Users/stanko/.cargo/bin:/Users/stanko/.kiex/elixirs/elixir-1.6.4/bin:/Users/stanko/.kiex/bin:/Users/stanko/.rbenv/shims:/Users/stanko/.rbenv/bin:/Users/stanko/.local/bin:/usr/local/opt/fzf/bin:/Users/stanko/.local/bin
PWD=/Users/stanko/Projects/Rust/public/alacritty
RBENV_ROOT=/Users/stanko/.rbenv
RBENV_SHELL=zsh
REBAR_PLT_DIR=/Users/stanko/.kerl/installs/20.3
SHELL=/usr/local/bin/zsh
SHLVL=2
SSH_AUTH_SOCK=/private/tmp/com.apple.launchd.HD9kfD1YnW/Listeners
TERM=screen-256color
TMPDIR=/var/folders/gk/vy9n6twd4pn613ffydfddw_r0000gn/T/
TMUX=/private/tmp/tmux-501/default,699,0
TMUX_PANE=%2
TMUX_PLUGIN_MANAGER_PATH=/Users/stanko/.tmux/plugins/
USER=stanko
USE_EDITOR=nvim
VISUAL=nvim
XPC_FLAGS=0x0
XPC_SERVICE_NAME=0
ZSH=/Users/stanko/.oh-my-zsh
_KERL_SAVED_REBAR_PLT_DIR=/Users/stanko/.kerl/installs/20.3
__CF_USER_TEXT_ENCODING=0x1F5:0x0:0x0
is_vim=echo "#{pane_current_command}" | grep -iqE "(^|\/)g?(view|n?vim?)(diff)?$"
_KERL_PATH_REMOVABLE=/Users/stanko/.kerl/installs/20.3/bin
_KERL_MANPATH_REMOVABLE=/Users/stanko/.kerl/installs/20.3/lib/erlang/man:/Users/stanko/.kerl/installs/20.3/man
_KERL_ACTIVE_DIR=/Users/stanko/.kerl/installs/20.3
_KERL_SAVED_ERL_AFLAGS=
_=/usr/bin/printenv
```

### PR DESCRIPTION
**Pull request by alexcrichton**
This commit disables usage of git templates whenever Cargo manages
repositories in its internal git database. Templates don't want to be
used at all in these situations and have been known to cause usability
bugs.

Closes #6240

### CODE REVIEW

**Comment by dwijnand** [comment]
Sounds good to me.

@bors: r+

File: tests/testsuite/git.rs [hunk 1]
```diff
@@ -2625,3 +2625,42 @@ fn use_the_cli() {
...
+            "#, git_project.url()),
+        ).file("src/main.rs", "fn main() {}")
+        .build();
+
+    File::create(paths::home().join(".gitconfig"))
+        .unwrap()
+        .write_all(
+            &format!(r#"
+                [init]
+                templatedir = {}
```
Comment (ehuss) [inline tests/testsuite/git.rs]:
```suggestion
                templatedir = "{}"
```

Comment (ehuss) [inline tests/testsuite/git.rs]:
```suggestion
                templatedir = "{}"
```

File: tests/testsuite/git.rs [hunk 2]
```diff
@@ -2625,3 +2625,42 @@ fn use_the_cli() {
...
+            "#, git_project2.url().to_file_path().unwrap().display()).as_bytes(),
```
Comment (ehuss) [inline tests/testsuite/git.rs]:
```suggestion
            "#,
                git_project2
                    .url()
                    .to_file_path()
                    .unwrap()
                    .display()
                    .to_string()
                    .replace("\\", "/")
            )
            .as_bytes(),
```

Comment (ehuss) [inline tests/testsuite/git.rs]:
```suggestion
            "#,
                git_project2
                    .url()
                    .to_file_path()
                    .unwrap()
                    .display()
                    .to_string()
                    .replace("\\", "/")
            )
            .as_bytes(),
```

**Comment by ehuss** [comment]
I think I broke/confused github trying the new suggestion feature.  Anyways, I was trying to say, the code needs to be changed to:

```
                templatedir = "{}"
            "#,
                git_project2
                    .url()
                    .to_file_path()
                    .unwrap()
                    .display()
                    .to_string()
                    .replace("\\", "/")
            )
            .as_bytes(),
```

For "C:/path/to/thing" for gitconfig style paths (double quotes and forward slashes).

**Pull request merged**